lib/backend/etcdv3/watcher.go

```diff
@@ -0,0 +1,203 @@
+
+import (
+	"context"
+	"strconv"
+
+	"github.com/coreos/etcd/clientv3"
+	log "github.com/sirupsen/logrus"
+
+	"github.com/projectcalico/libcalico-go/lib/backend/api"
+	"github.com/projectcalico/libcalico-go/lib/backend/model"
+)
+
+const (
+	outgoingBufSize = 100
+)
+
+// Watch entries in the datastore matching the resources specified by the ListInterface.
+func (c *etcdV3Client) Watch(l model.ListInterface, revision string) (api.WatchInterface, error) {
+	var rev int64
+	if len(revision) != 0 {
+		var err error
+		rev, err = strconv.ParseInt(revision, 10, 64)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	wc := &watcher{
+		client:     c,
+		list:       l,
+		initialRev: rev,
+		resultChan: make(chan api.WatchEvent, outgoingBufSize),
+		errChan:    make(chan error, 1),
+	}
+	wc.ctx, wc.cancel = context.WithCancel(context.Background())
+	go wc.run()
+	return wc, nil
+}
+
+// watcher implements watch.Interface.
+type watcher struct {
+	client     *etcdV3Client
+	initialRev int64
+	ctx        context.Context
+	cancel     context.CancelFunc
+	resultChan chan api.WatchEvent
+	errChan    chan error
+	list       model.ListInterface
+}
+
+func (wc *watcher) run() {
+	log.Info("Running watcher main loop")
+	watchClosedCh := make(chan struct{})
+	go wc.watchLoop(watchClosedCh)
+
+	select {
+	case err := <-wc.errChan:
+		log.Debug("watcher.run() loop received error event")
+		if err == context.Canceled {
+			log.Debug("error event was a cancel - ignoring")
+			break
+		}
+		errResult := &api.WatchEvent{
+			Type:  api.WatchError,
+			Error: err,
+		}
+
+		// Send the error result unless the user has cancelled.
+		select {
+		case wc.resultChan <- *errResult:
+		case <-wc.ctx.Done():
+		}
+	case <-watchClosedCh:
+		log.Debug("watcher.run() loop received watch closed event")
+	case <-wc.ctx.Done(): // user cancel
+		log.Debug("watcher.run() loop received done event")
+	}
+	log.Info("watcher.run() loop exiting")
+
+	// We use wc.ctx to reap all goroutines. Under whatever condition, we should stop them all.
+	// It's fine to double cancel.
+	wc.cancel()
+	close(wc.resultChan)
+}
+
+// Stop() implements the api.WatchInterface.
+// This calls through to the context cancel function.
+func (wc *watcher) Stop() {
+	wc.cancel()
+}
+
+// ResultChan() implements the api.WatchInterface.
+func (wc *watcher) ResultChan() <-chan api.WatchEvent {
+	return wc.resultChan
+}
+
+// listCurrent retrieves the existing entries and sends an event for each listed
+func (wc *watcher) listCurrent() error {
+	log.Info("Performing initial list with no revision")
+	list, err := wc.client.List(wc.list, "")
+	if err != nil {
+		return err
+	}
+
+	wc.initialRev, err = strconv.ParseInt(list.Revision, 10, 64)
+	if err != nil {
+		log.WithError(err).Error("List returned revision that could not be parsed")
+		return err
+	}
+
+	for _, kv := range list.KVPairs {
+		log.Info("Sending create events for each existing entry")
+		wc.sendEvent(&api.WatchEvent{
+			Type: api.WatchAdded,
```

Comment:
not sure I understand this part. Can you explain why all these events are `WatchAdded` type? Maybe even add a comment if it's not something obvious im not seeing

Reply:
Really just a copy of the k8s watch interface.  For the initial sync we List the current entries and send each current entry up as an event type.  Since these events exist - added is the event type of choice - doesn't make sense to use modified or deleted.

I'll add a comment.